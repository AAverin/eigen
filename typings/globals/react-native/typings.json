{
  "resolution": "main",
  "tree": {
    "src": "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/4c7d03b088d4e71cd9fe9c66ab7bd363049077c3/react-native/react-native.d.ts",
    "raw": "registry:dt/react-native#0.29.0+20160719232534",
    "typings": "https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/4c7d03b088d4e71cd9fe9c66ab7bd363049077c3/react-native/react-native.d.ts"
  }
}
