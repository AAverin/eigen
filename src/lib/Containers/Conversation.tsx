import * as moment from "moment"
import * as React from "react"
import * as Relay from "react-relay"

import { MetadataText, SmallHeadline } from "../Components/Inbox/Typography"

import { FlatList, ImageURISource, ViewProperties } from "react-native"
import ReversedFlatList from "react-native-reversed-flat-list"

import styled from "styled-components/native"
import colors from "../../data/colors"
import Composer from "../Components/Inbox/Conversations/Composer"
import Message from "../Components/Inbox/Conversations/Message"
import ArtworkPreview from "../Components/Inbox/Conversations/Preview/ArtworkPreview"

import ARSwitchBoard from "../NativeModules/SwitchBoard"

// tslint:disable-next-line:no-var-requires
const chevron: ImageURISource = require("../../../images/horizontal_chevron.png")

const Container = styled.View`
  flex: 1;
  flex-direction: column;
`
const Header = styled.View`
  align-self: stretch;
  margin-top: 10px;
  flex-direction: column;
  margin-bottom: 20px;
`

// This makes it really easy to style the HeaderTextContainer with space-between
const PlaceholderView = styled.View``

const HeaderTextContainer = styled.View`
  flexDirection: row
  justifyContent: space-between
`

const BackButtonPlaceholder = styled.Image`
  height: 12
  width: 7
  transform: rotate(180deg)
`

const DottedBorder = styled.View`
  height: 1
  borderWidth: 1
  borderStyle: dotted
  borderColor: ${colors["gray-regular"]}
  marginLeft: 20
  marginRight: 20

`

const MessagesList = styled(FlatList)`
  marginTop: 10
`

const ComposerContainer = styled.View`
  paddingTop: 5
  marginRight: 20
  marginLeft: 20
`

const PAGE_SIZE = 10

export class Conversation extends React.Component<RelayProps, any> {
  renderMessage({ item }) {
    // End of the list reached. Increase page size. Relay
    // will fetch only the required data to fill the new
    // page size.
    if (item.key === this.props.me.conversation.messages.edges.length - 1) {
      this.props.relay.setVariables({ pageSize: this.props.me.conversation.messages.edges.length + PAGE_SIZE })
    }
    const artwork = this.props.me.conversation.artworks[0]
    const conversation = this.props.me.conversation
    const partnerName = conversation.to.name
    return (
      <Message
        message={item}
        partnerName={partnerName}
        userName={conversation.from.name}
        artworkPreview={
          item.first_message &&
          <ArtworkPreview
            artwork={artwork}
            onSelected={() => ARSwitchBoard.presentNavigationViewController(this, artwork.href)}
          />
        }
      />
    )
  }

  render() {
    const conversation = this.props.me.conversation
    const partnerName = conversation.to.name
    const artwork = conversation.artworks[0]
    const messages = conversation.messages.edges.map((edge, index) => {
      return { first_message: index === 0, key: index, ...edge.node }
    })
    const lastMessage = messages[messages.length - 1]

    return (
      <Container>
        <Header>
          <HeaderTextContainer>
            <BackButtonPlaceholder source={chevron} />
            <SmallHeadline>
              {partnerName}
            </SmallHeadline>
            <PlaceholderView />
          </HeaderTextContainer>
        </Header>
        <ReversedFlatList
          data={messages}
          renderItem={this.renderMessage.bind(this)}
          length={messages.length}
          ItemSeparatorComponent={DottedBorder}
        />
        <ComposerContainer>
          <Composer
            onSubmit={text => {
              this.props.relay.commitUpdate(
                new SendConversationMessageMutation({
                  body_text: text,
                  reply_to_message_id: lastMessage.impulse_id,
                  conversation: this.props.me.conversation as any,
                })
              )

              /// Not sure how to get optimistic re-render?
              this.props.relay.setVariables({ pageSize: this.props.me.conversation.messages.edges.length + 1 })
            }}
          />
        </ComposerContainer>
      </Container>
    )
  }
}

interface MutationProps {
  body_text: string
  reply_to_message_id: string
  conversation: {
    __id: string
    id: string
    from: {
      email: string
    }
  }
}

class SendConversationMessageMutation extends Relay.Mutation<MutationProps, any> {
  static fragments = {
    conversation: () => Relay.QL`
      fragment on Conversation {
        __id
        id
        from {
          email
        }
      }
    `,
  }

  getMutation() {
    return Relay.QL`mutation { sendConversationMessage }`
  }

  getVariables() {
    return {
      id: this.props.conversation.id,
      from: this.props.conversation.from.email,
      body_text: this.props.body_text,
      reply_to_message_id: this.props.reply_to_message_id,
    }
  }

  getFatQuery() {
    return Relay.QL`
      fragment on SendConversationMessageMutationPayload {
        messageEdge
        conversation {
          messages
        }
      }
    `
  }

  getConfigs() {
    return [
      {
        type: "RANGE_ADD",
        parentName: "conversation",
        parentID: this.props.conversation.__id,
        connectionName: "messages",
        edgeName: "messageEdge",
        rangeBehaviors: {
          "": "append",
        },
      },
    ]
  }

}

export default Relay.createContainer(Conversation, {
  initialVariables: {
    pageSize: PAGE_SIZE,
    conversationID: null,
  },
  fragments: {
    me: () => Relay.QL`
      fragment on Me {
        conversation(id: $conversationID) {
          from {
            name
          }
          to {
            name
          }
          messages(first: $pageSize) {
            pageInfo {
              hasNextPage
            }
            edges {
              node {
                impulse_id
                ${Message.getFragment("message")}
              }
            }
          }
          artworks {
            href
            ${ArtworkPreview.getFragment("artwork")}
          }
          ${SendConversationMessageMutation.getFragment("conversation")}
        }
      }
    `,
  },
})

interface RelayProps {
  relay: any
  me: {
    conversation: {
      from: {
        name: string
      }
      to: {
        name: string
      }
      artworks: any[]
      messages: {
        pageInfo?: {
          hasNextPage: boolean
        }
        edges: Array<{
          node: { impulse_id: string } | null
        }>
      }
    }
  }
}
