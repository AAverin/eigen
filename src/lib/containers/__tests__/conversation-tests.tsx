import * as React from "react"
import "react-native"
import * as renderer from "react-test-renderer"
import Conversation from "../conversation"

it("looks correct when rendered", () => {
  const tree = renderer.create(<Conversation />).toJSON()
  expect(tree).toMatchSnapshot()
})
