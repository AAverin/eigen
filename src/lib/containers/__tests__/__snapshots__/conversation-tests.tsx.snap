// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`looks correct when rendered 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "marginLeft": 20,
        "marginRight": 20,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "stretch",
          "flexDirection": "column",
          "marginTop": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          undefined,
        ]
      }
    >
      <Image
        source={1}
        style={
          Array [
            Object {
              "height": 12,
              "transform": Array [
                Object {
                  "rotate": "180deg",
                },
              ],
              "width": 7,
            },
            undefined,
          ]
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 11,
              "textAlign": "left",
            },
            Object {},
            Object {
              "fontFamily": "Avant Garde Gothic ITCW01Dm",
            },
          ]
        }
      >
        PATRICK PARRISH GALLERY
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#cccccc",
              "fontSize": 11,
              "textAlign": "left",
            },
            Object {},
            Object {
              "fontFamily": "Avant Garde Gothic ITCW01Dm",
            },
          ]
        }
      >
        INFO
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "#e5e5e5",
            "borderStyle": "dotted",
            "borderWidth": 1,
            "height": 1,
            "marginTop": 30,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "alignSelf": "stretch",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "marginBottom": 10,
            "marginTop": 15,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#e5e5e5",
            "borderRadius": 20,
            "height": 20,
            "width": 20,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          Array [
            Object {
              "marginLeft": 10,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            Array [
              Object {
                "alignSelf": "stretch",
                "marginBottom": 10,
              },
              undefined,
            ],
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "black",
                "fontSize": 11,
                "textAlign": "left",
              },
              Array [
                Object {
                  "marginRight": 10,
                },
                undefined,
              ],
              Object {
                "fontFamily": "Avant Garde Gothic ITCW01Dm",
              },
            ]
          }
        >
          PATRICK PARRISH
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#cccccc",
                "fontSize": 11,
                "textAlign": "left",
              },
              Object {},
              Object {
                "fontFamily": "Avant Garde Gothic ITCW01Dm",
              },
            ]
          }
        >
          11:32AM
        </Text>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        Hi Katarina, thanks for reaching out with your interest in this great piece by Ian Stell.
      </Text>
    </View>
  </View>
  <View
    keyboardVerticalOffset={0}
    onLayout={[Function]}
    style={
      Array [
        undefined,
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      active={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#f8f8f8",
            "borderColor": "#e5e5e5",
            "borderRadius": 3,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 20,
          },
          undefined,
        ]
      }
    >
      <TextInput
        keyboardAppearance="dark"
        multiline={true}
        onChangeText={[Function]}
        onEndEditing={[Function]}
        onFocus={[Function]}
        placeholder="Reply..."
        placeholderTextColor="#666666"
        style={
          Object {
            "flex": 1,
            "fontFamily": "AGaramondPro-Regular",
            "fontSize": 12,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
          }
        }
      />
      <Text
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        allowFontScaling={true}
        containsText={false}
        disabled={false}
        ellipsizeMode="tail"
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "color": "#e5e5e5",
              "fontFamily": "Avant Garde Gothic ITCW01Dm",
              "fontSize": 12,
              "marginRight": 10,
            },
            undefined,
          ]
        }
        testID={undefined}
      >
        SEND
      </Text>
    </View>
  </View>
</View>
`;
